### PR TITLE
fix issue #15

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

**Comment by 9il** [comment]
ping @WebDrake

**Comment by WebDrake** [comment]
Looks OK in principle.  In some ways I'm not sure I'm comfortable with dual-purposing `max` as both a value and as a bitmask, but your call.

File: source/mir/random/engine/mersenne_twister.d [hunk 1]
```diff
@@ -30,8 +30,8 @@ struct MersenneTwisterEngine(UIntType, size_t w, size_t n, size_t m, size_t r,
     @disable this();
     @disable this(this);
 
-    private enum UIntType upperMask = ~((cast(UIntType) 1u << (UIntType.sizeof * 8 - (w - r))) - 1);
     private enum UIntType lowerMask = (cast(UIntType) 1u << r) - 1;
+    private enum UIntType upperMask = ~lowerMask;
```
Comment (WebDrake) [inline source/mir/random/engine/mersenne_twister.d]: Note here: according to https://en.wikipedia.org/wiki/Mersenne_Twister `upperMask` is supposed to be the lowest `w` bits of `~lowerMask`.

Comment (9il) [inline source/mir/random/engine/mersenne_twister.d]: Thanks! fixed in #17

**Comment by WebDrake** [comment]
Last comment I'll make on here: isn't the return value of `opCall` also supposed to be the lowest `w` bits of the generated value?  (It might be superfluous if all is set up correctly but my inclination is to be safe with this.)

**Comment by 9il** [comment]
Yes, but if `b` and `c` is correct, then `z` is correct too without `& max`

**Comment by WebDrake** [comment]
What are the constraints on `b` and `c` that would allow us to determine that they are correct in this respect?

**Comment by 9il** [comment]
They should be less then or equal to `max` (fit into lowest `w` bits)

**Comment by 9il** [comment]
the check is already here: `static assert(a <= max && b <= max && c <= max);`

**Comment by WebDrake** [comment]
OK, cool.  Thanks for the quick fixes -- I'll get them over into the phobos PR soon!

One remark -- I didn't include any constraints on `f` in my earlier PR but if I've understood correctly, the same principle applies: it should be less than or equal to `max`.

**Comment by 9il** [comment]
> One remark -- I didn't include any constraints on f in my earlier PR but if I've understood correctly, the same principle applies: it should be less than or equal to max.

Hmm, I am not sure. It will work with larger values too

**Comment by 9il** [comment]
see also https://github.com/libmir/mir-random/pull/18, it is required for ushort and ubyte types

**Comment by WebDrake** [comment]
> Hmm, I am not sure. It will work with larger values too

It appears to be applied as a compile-time constraint in the C++11 standard library version.  Try instantiating e.g.:

```C++
std::mersenne_twister_engine<
        UIntType, w, 624, 397, 31,
        0x9908b0df, 11, 0xffffffff, 7,
        0x9d2c5680, 15,
        0xefc60000, 18, (1uL << 32) - 1> gen;
```

... and then try instead:

```C++
std::mersenne_twister_engine<
        UIntType, w, 624, 397, 31,
        0x9908b0df, 11, 0xffffffff, 7,
        0x9d2c5680, 15,
        0xefc60000, 18, (1uL << 32)> gen;
```

**Pull request merged**